### PR TITLE
Adds support for i32x4.trunc_sat_f32x4_u

### PR DESCRIPTION
**Pull request by jlb6740**

### CODE REVIEW

**Comment by jlb6740** [comment]
Does not guard SIMD4.1 lowerings but a think other lowerings don't either so I was going to include that as a separate patchset that potentially did others as well.

**Review by cfallin — COMMENTED** [review]
Thanks for this -- these are quite subtle implementations! I have a few suggestions for the main explanation comment for the saturating conversion below, and I want to stare at the lowering a bit more as I'm not fully sure I grok it (I'm hoping the updated comment will help :-) ) but overall this seems reasonable.

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 1]
```diff
@@ -2776,7 +2776,78 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
...
+                    // so those get converted correctly as well. The tricky part lies converting
+                    // those positive values that lie inbetween the max unsigned positive value and
+                    // and the max signed positive value while saturating everything above the max
+                    // signed positive value. To do this properly we do the conversion with Cvttps2dq
+                    // again on a copy of the original pre-converted value (with negatives and NaN
+                    // already handled). Also, this second conversion is done on a value that subtracts
+                    // out the max signed int value. The result means when we add this converted vector
+                    // to the original converted vector we properly account for those values between
+                    // the max unsigned and max signed value. For those overflow outside the range of
+                    // the max unsigned range we create a mask to account for these lanes.
```
[the remark this answers is not in the document]
Comment (jlb6740) [reply cranelift/codegen/src/isa/x64/lower.rs:2792]: Cool .. good suggestion. I'll make it easier to follow and highlight some specific cases.

File: cranelift/codegen/src/isa/x64/lower.rs [hunk 2]
```diff
@@ -2776,7 +2776,78 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
...
+
+                    let tmp1 = ctx.alloc_tmp(RegClass::V128, types::I32X4);
```
[the remark this answers is not in the document]
Comment (jlb6740) [reply cranelift/codegen/src/isa/x64/lower.rs:2794]: Good point. I'll point some assertions in.

**Review by cfallin — APPROVED** [review]
This is now the best-documented FP lowering sequence in the backend -- thanks so much for the detailed comment! :-)

**Pull request merged**